plugin/plugin.go

```diff
@@ -4,22 +4,207 @@
+		Mirror         string   `envconfig:"PLUGIN_MIRROR"`                                         // Docker registry mirror
+		Insecure       bool     `envconfig:"PLUGIN_INSECURE"`                                       // Docker daemon enable insecure registries
+		StorageDriver  string   `envconfig:"PLUGIN_STORAGE_DRIVER"`                                 // Docker daemon storage driver
+		StoragePath    string   `envconfig:"PLUGIN_STORAGE_PATH" default:"/var/lib/docker"`         // Docker daemon storage path
+		Disabled       bool     `envconfig:"PLUGIN_DAEMON_OFF"`                                     // DOcker daemon is disabled (already running)
+		Debug          bool     `envconfig:"PLUGIN_DEBUG"`                                          // Docker daemon started in debug mode
+		Bip            string   `envconfig:"PLUGIN_BIP"`                                            // Docker daemon network bridge IP address
+		DNS            []string `envconfig:"PLUGIN_CUSTOM_DNS"`                                     // Docker daemon dns server
+		DNSSearch      []string `envconfig:"PLUGIN_CUSTOM_DNS_SEARCH"`                              // Docker daemon dns search domain
+		MTU            string   `envconfig:"PLUGIN_MTU"`                                            // Docker daemon mtu setting
+		IPv6           bool     `envconfig:"PLUGIN_IPV6"`                                           // Docker daemon IPv6 networking
+		Experimental   bool     `envconfig:"PLUGIN_EXPERIMENTAL"`                                   // Docker daemon enable experimental mode
+	}
+)
 
 // Args provides plugin execution arguments.
 type Args struct {
 	Pipeline
-
-	// Level defines the plugin log level.
-	Level string `envconfig:"PLUGIN_LOG_LEVEL"`
-
-	// TODO replace or remove
-	Param1 string `envconfig:"PLUGIN_PARAM1"`
-	Param2 string `envconfig:"PLUGIN_PARAM2"`
+	Login      Login
+	Daemon     Daemon
+	Level      string `envconfig:"PLUGIN_LOG_LEVEL"`
+	Dockerfile string `envconfig:"PLUGIN_DOCKERFILE" required:"true"`
+	Image      string `envconfig:"PLUGIN_IMAGE" required:"true"`
+	AuthToken  string `envconfig:"PLUGIN_SNYK"`
 }
 
 // Exec executes the plugin.
 func Exec(ctx context.Context, args Args) error {
-	// write code here
+	if !args.Daemon.Disabled {
+		startDaemon(args.Daemon)
+	}
+
+	// poll the docker daemon until it is started. This ensures the daemon is
+	// ready to accept connections before we proceed.
+	for i := 0; ; i++ {
+		cmd := commandInfo()
+		err := cmd.Run()
+		if err == nil {
+			break
+		}
+		if i == 15 {
+			fmt.Println("Unable to reach Docker Daemon after 15 attempts.")
+			break
+		}
+		time.Sleep(time.Second * 1)
+	}
+    // docker login
+	cmdL := commandLogin(args.Login)
+	err := cmdL.Run()
+	if err != nil {
+		return fmt.Errorf("error authenticating: %s", err)
+	}
+
+	fmt.Println("authentication successful")
+
+	var cmds []*exec.Cmd
+	if args.AuthToken != "" {
```

Comment:
I recommend adding some sort of debug logging if no token is provided (similar to what we do in the docker plugin) . This can help users diagnose plugin issues if they misconfigure a secret and it is not injected as expected.

Reply:
Cheers Brad, I've made that change.